__tests__/schema.sql

```diff
@@ -125,6 +125,49 @@ begin
 end;
 $$;
 ALTER FUNCTION graphile_worker.add_job(identifier text, payload json, queue_name text, run_at timestamp with time zone, max_attempts integer, job_key text, priority integer, flags text[]) OWNER TO graphile_worker_role;
+CREATE FUNCTION graphile_worker.complete_batch(worker_id text, success_ids bigint[], failures json[]) RETURNS void
+    LANGUAGE plpgsql STRICT
+    AS $$
+declare
+  v_row "graphile_worker".jobs;
+begin
+  if array_length(failures, 1) is not null then
+    with fail_jobs as (
+      update "graphile_worker".jobs
+        set
+          last_error = f->>'message',
+          run_at = greatest(now(), run_at) + (exp(least(attempts, 10))::text || ' seconds')::interval,
+          locked_by = null,
+          locked_at = null
+        from unnest(failures) f
+        where id = (f->>'id')::bigint and locked_by = worker_id
+        returning queue_name
+    )
+    update "graphile_worker".job_queues jq
+      set locked_by = null, locked_at = null
+      where exists (
+        select 1
+        from fail_jobs f
+        where jq.queue_name = f.queue_name
+      ) and locked_by = worker_id;
+  end if;
+  if array_length(success_ids, 1) is not null then
+    with success_jobs as (
+      delete from "graphile_worker".jobs
+        where id = any(success_ids)
+      returning queue_name
+    )
+    update "graphile_worker".job_queues jq
+      set locked_by = null, locked_at = null
+      where exists (
+        select 1
+        from success_jobs s
+        where jq.queue_name = s.queue_name
+      ) and locked_by = worker_id;
+  end if;
+end;
+$$;
```

Comment:
If you drop the `if` statements you could convert this to `sql` rather than `plpgsql` which might have an impact on performance. Maybe.

Reply:
I think this is a good idea anyway, always better to use plain sql if possible, imo